Implement 95% of the NO_PERL functionality, to build Git without any Perl
support, because some Gentoo users want a Git without any Perl whatesoever
(Gentoo bug #214168).

Remaining bits are doing configure.ac as well as git-remote usage in:
t5502-quickfetch.sh
t5512-ls-remote.sh
t5520-pull.sh

Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>
Bugzilla-URL: http://bugs.gentoo.org/show_bug.cgi?id=214168
Notes: Ported from 20080423-git-1.5.5.1-noperl.patch
Notes: Ported from 20080322-git-1.5.4.5-noperl.patch
Notes: Ported from 20080528-git-1.5.6.1-noperl.patch

diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/builtin-add.c git-1.5.6.1/builtin-add.c
--- git-1.5.6.1.orig/builtin-add.c	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/builtin-add.c	2008-05-28 10:38:12.566970120 -0700
@@ -135,6 +135,7 @@
         free(seen);
 }
 
+#ifndef NO_PERL
 static const char **validate_pathspec(int argc, const char **argv, const char *prefix)
 {
 	const char **pathspec = get_pathspec(prefix, argv);
@@ -170,6 +171,7 @@
 	free(args);
 	return status;
 }
+#endif
 
 static struct lock_file lock_file;
 
@@ -182,8 +184,10 @@
 	OPT__DRY_RUN(&show_only),
 	OPT__VERBOSE(&verbose),
 	OPT_GROUP(""),
+#ifndef NO_PERL
 	OPT_BOOLEAN('i', "interactive", &add_interactive, "interactive picking"),
 	OPT_BOOLEAN('p', "patch", &patch_interactive, "interactive patching"),
+#endif
 	OPT_BOOLEAN('f', "force", &ignored_too, "allow adding otherwise ignored files"),
 	OPT_BOOLEAN('u', "update", &take_worktree_changes, "update tracked files"),
 	OPT_BOOLEAN( 0 , "refresh", &refresh_only, "don't add, only refresh the index"),
@@ -198,10 +202,12 @@
 
 	argc = parse_options(argc, argv, builtin_add_options,
 			  builtin_add_usage, 0);
+#ifndef NO_PERL
 	if (patch_interactive)
 		add_interactive = 1;
 	if (add_interactive)
 		exit(interactive_add(argc, argv, prefix));
+#endif
 
 	git_config(git_default_config);
 
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/builtin-commit.c git-1.5.6.1/builtin-commit.c
--- git-1.5.6.1.orig/builtin-commit.c	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/builtin-commit.c	2008-05-28 10:38:25.556703720 -0700
@@ -97,7 +97,9 @@
 	OPT_GROUP("Commit contents options"),
 	OPT_BOOLEAN('a', "all", &all, "commit all changed files"),
 	OPT_BOOLEAN('i', "include", &also, "add specified files to index for commit"),
+#ifndef NO_PERL
 	OPT_BOOLEAN(0, "interactive", &interactive, "interactively add files"),
+#endif
 	OPT_BOOLEAN('o', "only", &only, "commit only specified files"),
 	OPT_BOOLEAN('n', "no-verify", &no_verify, "bypass pre-commit hook"),
 	OPT_BOOLEAN(0, "amend", &amend, "amend previous commit"),
@@ -217,6 +219,7 @@
 	struct path_list partial;
 	const char **pathspec = NULL;
 
+#ifndef NO_PERL
 	if (interactive) {
 		interactive_add(argc, argv, prefix);
 		if (read_cache() < 0)
@@ -224,6 +227,7 @@
 		commit_style = COMMIT_AS_IS;
 		return get_index_file();
 	}
+#endif
 
 	if (read_cache() < 0)
 		die("index file corrupt");
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/Makefile git-1.5.6.1/Makefile
--- git-1.5.6.1.orig/Makefile	2008-05-28 10:38:03.608016000 -0700
+++ git-1.5.6.1/Makefile	2008-05-28 10:38:12.571288044 -0700
@@ -130,6 +130,8 @@
 #
 # Define NO_PERL_MAKEMAKER if you cannot use Makefiles generated by perl's
 # MakeMaker (e.g. using ActiveState under Cygwin).
+
+# Define NO_PERL if you do not want Perl scripts at all.
 #
 # Define NO_TCLTK if you do not want Tcl/Tk GUI.
 #
@@ -255,6 +257,8 @@
 SCRIPT_SH += git-submodule.sh
 SCRIPT_SH += git-web--browse.sh
 
+SCRIPTS = $(patsubst %.sh,%,$(SCRIPT_SH))
+ifndef NO_PERL
 SCRIPT_PERL += git-add--interactive.perl
 SCRIPT_PERL += git-archimport.perl
 SCRIPT_PERL += git-cvsexportcommit.perl
@@ -263,10 +267,11 @@
 SCRIPT_PERL += git-relink.perl
 SCRIPT_PERL += git-send-email.perl
 SCRIPT_PERL += git-svn.perl
-
-SCRIPTS = $(patsubst %.sh,%,$(SCRIPT_SH)) \
-	  $(patsubst %.perl,%,$(SCRIPT_PERL)) \
-	  git-instaweb
+SCRIPTS += $(patsubst %.perl,%,$(SCRIPT_PERL)) \
+			git-instaweb
+else
+SCRIPT_PERL =
+endif
 
 # Empty...
 EXTRA_PROGRAMS =
@@ -315,7 +320,10 @@
 ALL_PROGRAMS = $(PROGRAMS) $(SCRIPTS)
 
 # what 'all' will build but not install in gitexecdir
-OTHER_PROGRAMS = git$X gitweb/gitweb.cgi
+OTHER_PROGRAMS = git$X
+ifndef NO_PERL
+OTHER_PROGRAMS += gitweb/gitweb.cgi
+endif
 
 # Set paths to tools early so that they can be used for version tests.
 ifndef SHELL_PATH
@@ -716,6 +724,10 @@
 	endif
 endif
 
+ifdef NO_PERL
+	BASIC_CFLAGS += -DNO_PERL
+endif
+
 ifdef ZLIB_PATH
 	BASIC_CFLAGS += -I$(ZLIB_PATH)/include
 	EXTLIBS += -L$(ZLIB_PATH)/$(lib) $(CC_LD_DYNPATH)$(ZLIB_PATH)/$(lib)
@@ -896,6 +908,11 @@
 ifeq ($(TCLTK_PATH),)
 NO_TCLTK=NoThanks
 endif
+ifeq ($(PERL_PATH),)
+NO_PERL=NoThanks
+export NO_PERL
+export NO_PERL_MAKEMAKER
+endif
 
 QUIET_SUBDIR0  = +$(MAKE) -C # space to separate -C and subdir
 QUIET_SUBDIR1  =
@@ -968,7 +985,9 @@
 	$(QUIET_SUBDIR0)git-gui $(QUIET_SUBDIR1) all
 	$(QUIET_SUBDIR0)gitk-git $(QUIET_SUBDIR1) all
 endif
+ifndef NO_PERL
 	$(QUIET_SUBDIR0)perl $(QUIET_SUBDIR1) PERL_PATH='$(PERL_PATH_SQ)' prefix='$(prefix_SQ)' all
+endif
 	$(QUIET_SUBDIR0)templates $(QUIET_SUBDIR1)
 
 strip: $(PROGRAMS) git$X
@@ -1007,6 +1026,7 @@
 	chmod +x $@+ && \
 	mv $@+ $@
 
+ifndef NO_PERL
 $(patsubst %.perl,%,$(SCRIPT_PERL)): perl/perl.mak
 
 perl/perl.mak: GIT-CFLAGS perl/Makefile perl/Makefile.PL
@@ -1065,6 +1085,7 @@
 	    $@.sh > $@+ && \
 	chmod +x $@+ && \
 	mv $@+ $@
+endif # NO_PERL
 
 configure: configure.ac
 	$(QUIET_GEN)$(RM) $@ $<+ && \
@@ -1218,7 +1239,9 @@
 	$(INSTALL) $(ALL_PROGRAMS) '$(DESTDIR_SQ)$(gitexecdir_SQ)'
 	$(INSTALL) git$X '$(DESTDIR_SQ)$(bindir_SQ)'
 	$(MAKE) -C templates DESTDIR='$(DESTDIR_SQ)' install
+ifndef NO_PERL
 	$(MAKE) -C perl prefix='$(prefix_SQ)' DESTDIR='$(DESTDIR_SQ)' install
+endif
 ifndef NO_TCLTK
 	$(MAKE) -C gitk-git install
 	$(MAKE) -C git-gui install
@@ -1307,9 +1330,11 @@
 	$(RM) -r $(GIT_TARNAME) .doc-tmp-dir
 	$(RM) $(GIT_TARNAME).tar.gz git-core_$(GIT_VERSION)-*.tar.gz
 	$(RM) $(htmldocs).tar.gz $(manpages).tar.gz
-	$(RM) gitweb/gitweb.cgi
 	$(MAKE) -C Documentation/ clean
+ifndef NO_PERL
+	$(RM) gitweb/gitweb.cgi
 	$(MAKE) -C perl clean
+endif
 	$(MAKE) -C templates/ clean
 	$(MAKE) -C t/ clean
 ifndef NO_TCLTK
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/lib-git-svn.sh git-1.5.6.1/t/lib-git-svn.sh
--- git-1.5.6.1.orig/t/lib-git-svn.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/lib-git-svn.sh	2008-05-28 10:38:12.571288044 -0700
@@ -6,6 +6,12 @@
 	test_done
 	exit
 fi
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping git-svn tests, NO_PERL defined' :
+	test_done
+	exit
+fi
 
 GIT_DIR=$PWD/.git
 GIT_SVN_DIR=$GIT_DIR/svn/git-svn
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t5505-remote.sh git-1.5.6.1/t/t5505-remote.sh
--- git-1.5.6.1.orig/t/t5505-remote.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t5505-remote.sh	2008-05-28 10:38:12.571288044 -0700
@@ -3,6 +3,12 @@
 test_description='git remote porcelain-ish'
 
 . ./test-lib.sh
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping git-cvsimport tests, NO_PERL defined' :
+	test_done
+	exit
+fi
 
 setup_repository () {
 	mkdir "$1" && (
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t7501-commit.sh git-1.5.6.1/t/t7501-commit.sh
--- git-1.5.6.1.orig/t/t7501-commit.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t7501-commit.sh	2008-05-28 10:38:12.571288044 -0700
@@ -38,7 +38,7 @@
 	"echo King of the bongo >file &&
 	! git-commit -m foo -a file"
 
-test_expect_success \
+[ -z "$NO_PERL" ] && test_expect_success \
 	"using paths with --interactive" \
 	"echo bong-o-bong >file &&
 	! echo 7 | git-commit -m foo --interactive file"
@@ -119,7 +119,7 @@
 	"echo 'gak' >file && \
 	 git-commit -m 'author' --author 'Rubber Duck <rduck@convoy.org>' -a"
 
-test_expect_success \
+[ -z "$NO_PERL" ] && test_expect_success \
 	"interactive add" \
 	"echo 7 | git-commit --interactive | grep 'What now'"
 
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t9001-send-email.sh git-1.5.6.1/t/t9001-send-email.sh
--- git-1.5.6.1.orig/t/t9001-send-email.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t9001-send-email.sh	2008-05-28 10:38:12.571288044 -0700
@@ -2,6 +2,12 @@
 
 test_description='git-send-email'
 . ./test-lib.sh
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping git-send-email tests, NO_PERL defined' :
+	test_done
+	exit
+fi
 
 PROG='git send-email'
 test_expect_success \
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t9200-git-cvsexportcommit.sh git-1.5.6.1/t/t9200-git-cvsexportcommit.sh
--- git-1.5.6.1.orig/t/t9200-git-cvsexportcommit.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t9200-git-cvsexportcommit.sh	2008-05-28 10:38:12.571288044 -0700
@@ -13,6 +13,12 @@
     test_done
     exit
 fi
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping git-cvsexportcommit tests, NO_PERL defined' :
+	test_done
+	exit
+fi
 
 CVSROOT=$(pwd)/cvsroot
 CVSWORK=$(pwd)/cvswork
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t9400-git-cvsserver-server.sh git-1.5.6.1/t/t9400-git-cvsserver-server.sh
--- git-1.5.6.1.orig/t/t9400-git-cvsserver-server.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t9400-git-cvsserver-server.sh	2008-05-28 10:38:12.573745493 -0700
@@ -17,6 +17,12 @@
     test_done
     exit
 fi
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping git-cvsserver tests, NO_PERL defined' :
+	test_done
+	exit
+fi
 perl -e 'use DBI; use DBD::SQLite' >/dev/null 2>&1 || {
     test_expect_success 'skipping git-cvsserver tests, Perl SQLite interface unavailable' :
     test_done
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t9500-gitweb-standalone-no-errors.sh git-1.5.6.1/t/t9500-gitweb-standalone-no-errors.sh
--- git-1.5.6.1.orig/t/t9500-gitweb-standalone-no-errors.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t9500-gitweb-standalone-no-errors.sh	2008-05-28 10:38:12.573745493 -0700
@@ -67,6 +67,13 @@
 }
 
 . ./test-lib.sh
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping gitweb-standalone-no-errors tests, NO_PERL defined' :
+	test_done
+	exit
+fi
+
 
 perl -MEncode -e 'decode_utf8("", Encode::FB_CROAK)' >/dev/null 2>&1 || {
     test_expect_success 'skipping gitweb tests, perl version is too old' :
diff -Nuar --exclude '*.orig' --exclude '*.rej' git-1.5.6.1.orig/t/t9600-cvsimport.sh git-1.5.6.1/t/t9600-cvsimport.sh
--- git-1.5.6.1.orig/t/t9600-cvsimport.sh	2008-05-28 00:56:46.000000000 -0700
+++ git-1.5.6.1/t/t9600-cvsimport.sh	2008-05-28 10:38:12.573745493 -0700
@@ -15,6 +15,12 @@
 	test_done
 	exit
 fi
+if test -n "$NO_PERL"
+then
+	test_expect_success 'skipping git-cvsimport tests, NO_PERL defined' :
+	test_done
+	exit
+fi
 
 cvsps_version=`cvsps -h 2>&1 | sed -ne 's/cvsps version //p'`
 case "$cvsps_version" in
